Node distribution (#70)
AAU-Dat/P9-ZKP: report/src/sections/04-experimental-protocol.tex

\section{Experimental protocol}\label{sec:experimental-protocol}
In this section we will cover the motivation and the workings of the attack that will be performed.
We also describe the setup of the experiment that will be conducted as well as the experiment itself.

\subsection{Motivation}\label{subsec:proposer-dos-attack}
\subsubsection{Proposer DoS Attack}

> In this section, we will be describing the attack that we will be using as a basis for our experiment in~\autoref{sec:experimental-protocol}.

Reviewer: ```suggestion
In this section, we will be describing the attack that acts as motivation for doing our experiment.
```

that are now available because of the attack.

\subsection{Attack description}\label{subsec:attack-description}
The overall goal of the attack is to be able to withhold validators from performing their eventual proposer duties as mentioned in~\cite{EthereumAttackDefense2024}.
The idea is that the adversary would interrupt validators using a~\gls{dos} attack.
The attack is divided in two parts.

> First being that we need information regarding two things; when a specific validator is going to produce a block, and what the IP address of the validator is.

Reviewer: ```suggestion
First we need to know validator IP addresses, found by de-anonymizing the validators.
```

This part of the attack will be explored below.

> The second part of the attack is to perform the~\gls{dos} attack on the de-anonymized validators.

Reviewer: ```suggestion
The second part of the attack is to find out when a specific validator is going to produce a block.
Utilizing the de-anonymized validators, the adversary should then be able to perform a~\gls{dos} attack on the upcoming block proposers.
```

\subsubsection{Getting validator IPs}
When a node is a part of the Ethereum network,
it receives messages in the form of attestations from other nodes in the network.
These attestations are split into two different categories, backbone and non-backbone attestations.
A backbone attestation is typically a routine message sent between peers in the same subnet.
This means that the nodes are simply forwarding messages to each other that they did not construct themselves.
To explain non-backbone attestations,
we refer directly to the observation found in ~\cite{heimbach2024deanonymizingethereumvalidatorsp2p}.
\\\\
\textit{\textbf{Observation 1.} An ideal peer will only send us an attestation in a subnet they are not a backbone of
if they are the signer of the attestation,
     and we are in their fanout for the corresponding subnet of the attestation.}
\\\\

So optimally, all non-backbone attestations that our node receives should be constructed by the node that sent it to us.
Therefore, in a perfect scenario,
we would couple the validator, associated with the non-backbone attestation, to the peer that sent the message.

> Though, there is a lot of things that are uncertain when sending messages over the internet.

Reviewer: ```suggestion
Though, there are a lot of things that are uncertain when sending messages over the internet.
```

So having matched the upcoming proposers with de-anonymized validators,
an adversary now ideally knows the IP-addresses of several upcoming proposers.

This allows for the adversary to~\gls{dos} the specific IP-address,
which could leave the proposer inactive at their designated slot.

\subsection{DoS attack implementation}\label{subsec:attack-implementation}
the work in this paper revolves around modifying a fork of the Prysm consensus client.

To be able to implement the attack,
it is a case of finding where the different functionalities are written in the code.
It is enough to modify just two classes in the code to reach a point where the attack is possible.
Two entities are logged, peers and attestations.
This translates almost directly to the code in the Prysm client, as each of these can be handled by a class each.

> It is also an important thing to keep in mind is that the core functionality of Prysm is never changed.

Reviewer: ```suggestion
It is also an important thing to keep in mind that the core functionality of Prysm is never changed.
```
